runtime/src/main/cpp/ObjCExport.mm

```diff
@@ -112,21 +112,13 @@ inline static OBJ_GETTER(AllocInstanceWithAssociatedObject, const TypeInfo* type
 static void initializeClass(Class clazz);
 extern "C" ALWAYS_INLINE void Kotlin_ObjCExport_releaseAssociatedObject(void* associatedObject);
 
-static inline id AtomicSetAssociatedObject(ObjHeader* obj, id associatedObject) {
-  if (!PermanentOrFrozen(obj)) {
-    SetAssociatedObject(obj, associatedObject);
-    return associatedObject;
-  } else {
-    void* old = __sync_val_compare_and_swap(&obj->meta_object()->associatedObject_, nullptr, (void*)associatedObject);
-    if (old == nullptr) {
-      return associatedObject;
-    } else {
-      Kotlin_ObjCExport_releaseAssociatedObject((void*)associatedObject);
-      return (id)old;
-    }
-  }
+static inline id AtomicCompareAndSwapAssociatedObject(ObjHeader* obj, id expectedValue, id newValue) {
+  return (id)__sync_val_compare_and_swap(&obj->meta_object()->associatedObject_, (void*)expectedValue, (void*)newValue);
```

Comment:
Maybe `reinterpret_cast<id>` here?

Reply:
After taking a fresh look, I consider turning this into 
```
id* location = reinterpret_cast<id*>(&obj->meta_object()->associatedObject_);
return __sync_val_compare_and_swap(location, expectedValue, newValue);
```
WDYT?